Compression/Compression.cs

```diff
@@ -524,6 +527,91 @@ public static StreamReader Unzip(string filename, string zipEntryName, out ZipFi
             return reader;
         }
 
+
+        // ZipArchive cache used by UnzipCached method
+        private static ConcurrentDictionary<string, ZipArchive> _zipArchiveCache = new ConcurrentDictionary<string, ZipArchive>();
+
+        /// <summary>
+        /// Clears ZipArchive cached used by UnzipCached method
+        /// </summary>
+        public static void ClearZipArchiveCache()
+        {
+            foreach (var zip in _zipArchiveCache)
+            {
+                zip.Value.Dispose();
+            }
+
+            _zipArchiveCache.Clear();
+        }
+
+        /// <summary>
+        /// Streams a local zip file using a streamreader and cached weak reference to ZipArchive in order to optimize method performance.
+        /// Important: the caller must call Dispose() on the returned ZipFile instance.
+        /// </summary>
+        /// <param name="filename">Location of the original zip file</param>
+        /// <param name="zipEntryName">The zip entry name to open a reader for. Specify null to access the first entry</param>
+        /// <param name="zipOut">The ZipArchive instance to be returned to the caller</param>
+        /// <returns>Stream reader of the first file contents in the zip file</returns>
+        public static StreamReader UnzipCached(string filename, string zipEntryName, out ZipArchive outZip)
+        {
+            StreamReader reader = null;
+            ZipArchive zip = null;
+            outZip = null;
+
+            try
+            {
+                if (File.Exists(filename))
+                {
+                    try
+                    {
+                        if (!_zipArchiveCache.TryGetValue(filename, out zip))
+                        {
+                            var file = File.OpenRead(filename);
+                            // reading up first two bytes 
+                            // http://george.chiramattel.com/blog/2007/09/deflatestream-block-length-does-not-match.html
+                            file.ReadByte(); file.ReadByte();
+
+                            zip = new ZipArchive(file);
+                            _zipArchiveCache[filename] = zip;
+                        }
+
+                        lock (_zipArchiveCache[filename])
```

Comment:
This piece is victim of performance optimization experiments, and may not look nice. Will add a todo to refactor it.